File: WePeiYang/lostfound/src/main/java/com/twtstudio/tjwhm/lostfound/waterfall/WaterfallFragment.java
package com.twtstudio.tjwhm.lostfound.waterfall;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.twtstudio.tjwhm.lostfound.R;
import com.twtstudio.tjwhm.lostfound.R2;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.facebook.common.internal.Ints.max;

/**
 * Created by tjwhm on 2017/7/3.
 **/

public class WaterfallFragment extends Fragment implements WaterfallContract.WaterfallView {
    @BindView(R2.id.waterfall_refresh)
    SwipeRefreshLayout water_refresh;
    @BindView(R2.id.waterfall_recyclerView)
    RecyclerView waterfall_recyclerView;
    @BindView(R2.id.waterfall_no_res)
    LinearLayout waterfall_no_res;
    private WaterfallTableAdapter tableAdapter;
    private StaggeredGridLayoutManager layoutManager = new StaggeredGridLayoutManager(2, LinearLayoutManager.VERTICAL);
    private WaterfallBean waterfallBean = new WaterfallBean();
    private int page = 1;
    private boolean isLoading = false;
    private boolean isRefresh = false;
    String lostOrFound;
    int type = -1;

    private WaterfallContract.WaterfallPresenter waterfallPresenter = new WaterfallPresenterImpl(this);

    public static WaterfallFragment newInstance(String type) {

        Bundle args = new Bundle();
        args.putString("index", type);
        WaterfallFragment fragment = new WaterfallFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.lf_fragment_waterfall, container, false);
        ButterKnife.bind(this, view);
        waterfall_recyclerView.setLayoutManager(layoutManager);
        waterfallBean.data = new ArrayList<>();
        waterfall_no_res.setVisibility(View.GONE);
        Bundle bundle = getArguments();
        lostOrFound = bundle.getString("index");
        tableAdapter = new WaterfallTableAdapter(waterfallBean, getActivity(), lostOrFound);
        waterfall_recyclerView.setAdapter(tableAdapter);
        water_refresh.setOnRefreshListener(this::refresh);


        waterfall_recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                int totalCount = layoutManager.getItemCount();
                int[] lastPositions = new int[layoutManager.getSpanCount()];
                layoutManager.findLastCompletelyVisibleItemPositions(lastPositions);

                int lastPosition = max(lastPositions);
                if (!isLoading && totalCount < lastPosition + 2 && lastPosition != -1) {
                    if (type == -1) {
                        ++page;
                        isLoading = true;
                        waterfallPresenter.loadWaterfallData(lostOrFound, page);
                    } else {
                        ++page;
                        isLoading = true;
                        waterfallPresenter.loadWaterfallDataWithType(lostOrFound, page, type);
                    }
                }
            }
        });
        return view;
    }

    @Override
    public void setWaterfallData(WaterfallBean waterfallBean) {
        if (waterfallBean.data.size() == 0 && page == 1) {
            waterfall_no_res.setVisibility(View.VISIBLE);
        } else {
            waterfall_no_res.setVisibility(View.GONE);
        }
        this.waterfallBean.error_code = waterfallBean.error_code;
        this.waterfallBean.message = waterfallBean.message;
        if (isRefresh) {
            this.waterfallBean.data.clear();
        }
        this.waterfallBean.data.addAll(waterfallBean.data);
        tableAdapter.notifyDataSetChanged();
        water_refresh.setRefreshing(false);
        isLoading = false;
        isRefresh = false;
    }

    @Override
    public void loadWaterfallDataWithType(int type) {
        this.type = type;
        page = 1;
        isRefresh = true;
        waterfallPresenter.loadWaterfallDataWithType(lostOrFound, page, type);
    }

    @Override
    public void onResume() {
        super.onResume();
        refresh();
    }

    private void refresh() {
        isLoading = true;
        isRefresh = true;
        page = 1;
        waterfallPresenter.loadWaterfallData(lostOrFound, page);
        type = -1;
    }
}

File: WePeiYang/lostfound2/src/main/java/com/example/lostfond2/waterfall/WaterfallFragment.kt
package com.example.lostfond2.waterfall

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import com.example.lostfond2.R
import com.example.lostfond2.service.MyListDataOrSearchBean
import kotlinx.android.synthetic.main.lf_fragment_waterfall.*

class WaterfallFragment : Fragment(), WaterfallContract.WaterfallView {

    private lateinit var tableAdapter: WaterfallTableAdapter
    private val layoutManager = StaggeredGridLayoutManager(2, LinearLayoutManager.VERTICAL)
    private var isLoading = false
    private var isRefresh = false
    var beanList = ArrayList<MyListDataOrSearchBean>()
    var lostOrFound = "lost"
    var type = -1
    var page = 1
    private val waterfallPresenter = WaterfallPresenterImpl(this)

    companion object {
        fun newInstance(type: String): WaterfallFragment {
            val args = Bundle()
            args.putString("index", type)
            val fragment = WaterfallFragment()
            fragment.arguments = args

            return fragment
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.lf_fragment_waterfall, container, false)
        val waterfall_refresh = view.findViewById<SwipeRefreshLayout>(R.id.waterfall_refresh)
        val waterfall_recyclerView = view.findViewById<RecyclerView>(R.id.waterfall_recyclerView)
        val waterfall_no_res = view.findViewById<LinearLayout>(R.id.waterfall_no_res)

        waterfall_recyclerView.layoutManager = layoutManager
        waterfall_no_res.visibility = View.GONE
        val bundle = arguments
        lostOrFound = bundle!!.getString("index")
        tableAdapter = WaterfallTableAdapter(beanList, this.activity!!, lostOrFound)
        waterfall_recyclerView.adapter = tableAdapter
        waterfall_refresh.setOnRefreshListener(this::refresh)

        waterfall_recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val totalCount = layoutManager.itemCount
                var lastPositions: IntArray = IntArray(layoutManager.spanCount)
                layoutManager.findLastCompletelyVisibleItemPositions(lastPositions)
                val lastPosition = lastPositions.max()

                if (!isLoading && (totalCount < lastPosition!! + 2) && lastPosition != -1) {
                    page++
                    isLoading = true

                    waterfallPresenter.apply {
                        if (type == -1) {
                            loadWaterfallData(lostOrFound, page)
                        } else {
                            loadWaterfallDataWithType(lostOrFound, page, type)
                        }
                    }
                }
            }
        })

        return view
    }

    override fun setWaterfallData(newBeanList: List<MyListDataOrSearchBean>) {
        waterfall_no_res.apply {
            if (newBeanList.size == 0 && page == 1) {
                visibility = View.VISIBLE
            } else {
                visibility = View.GONE
            }

            if (isRefresh) {
                beanList.clear()
            }

            beanList.addAll(newBeanList)
            tableAdapter.notifyDataSetChanged()
            waterfall_refresh.isRefreshing = false
            isLoading = false
            isRefresh = false
        }
    }

    override fun loadWaterfallDataWithType(type: Int) {
        this.type = type
        page = 1
        isRefresh = true
        waterfallPresenter.loadWaterfallDataWithType(lostOrFound, page, type)
    }

    override fun onResume() {
        super.onResume()
        waterfallPresenter.loadWaterfallData(lostOrFound,type)
    }

    private fun refresh() {
        isLoading = true
        isRefresh = true
        page = 1
        type = -1
        waterfallPresenter.loadWaterfallData(lostOrFound, page)
    }
}